File: content/puzzle/aoc/25/d10.hs
module Main where

import Data.Bits (shiftL, xor, (.|.))
import qualified Data.IntSet as IS
import qualified Data.List as L
import Data.Ratio ((%), denominator)

type Mask = Int
type Button = [Int]
type Constraint = ([Rational], Rational) -- c · x <= b

parseLine :: String -> (String, [Button], [Int])
parseLine s = (patternStr, buttons, targets)
  where
    -- pattern between [ ]
    (patternStr, rest) = span (/= ']') . tail $ dropWhile (/= '[') s
    -- buttons between ( )
    buttons = goButtons rest
    goButtons :: String -> [Button]
    goButtons str =
      case dropWhile (/= '(') str of
        "" -> []
        (_ : xs) ->
          let (nums, rest) = span (/= ')') xs
           in parseInts nums : goButtons (drop 1 rest)
    -- targets between { }
    targets =
      let inside = takeWhile (/= '}') . tail $ dropWhile (/= '{') s
       in parseInts inside

parseInts :: String -> [Int]
parseInts = map read . words . map (\c -> if c == ',' then ' ' else c)

-- part 1: BFS over light states (<= 2^10 states)
-- XOR is communicative and associative, so we only really care about pressing the button 0 or 1 time.
-- hence this is a bitmask operations.
-- state space is 2^n possible light configurations (n = number of lights)
--
-- Given that we run from state 0 (all off), each edge would be pressing the button, which XORs the current state with given button's mask.
patternMask :: String -> Mask
patternMask = go 0 0
  where
    go _ acc [] = acc
    go i acc (c : cs) =
      let bit = if c == '#' then 1 else 0
       in go (i + 1) (acc .|. (bit `shiftL` i)) cs

buttonMask :: Button -> Mask
buttonMask = foldl (\m i -> m `xor` (1 `shiftL` i)) 0

minPressLights :: Mask -> [Mask] -> Int
minPressLights target masks = bfs (IS.singleton 0) [(0, 0)]
  where
    bfs _ [] = error "unreachable"
    bfs seen ((state, d) : q)
      | state == target = d
      | otherwise =
          let nexts =
                [ state' | m <- masks, let state' = state `xor` m, not (IS.member state' seen)
                ]
              seen' = foldl (flip IS.insert) seen nexts
              q' = q ++ zip nexts (repeat (d + 1))
           in bfs seen' q'

-- part 2: we can frame it as minimize sum(x_i) subject to Ax = b with x >= 0 integer
--
-- A[i][j] = 1 if button j affects counter i, else 0
-- b[j] = target value for counter i
-- x[j] = number of time to press button j
--
-- Here, we use Gauss elimination for RREF [A|b], such that each pivot row gives x_pivot = b_i - sum(coeff * x_free). If the system is fully determined then we have an unique solution, otherwise we brute-force the bounded space to find the miminum total presses.
gaussJordan :: [[Rational]] -> [Rational] -> ([[Rational]], [Rational], [Int])
gaussJordan a b = go 0 0 a b []
  where
    rows = length a
    cols = length (head a)

    go _ _ mat rhs pivots | null mat = ([], [], pivots)
    go r c mat rhs pivots
      | r >= rows || c >= cols = (mat, rhs, pivots)
      | otherwise =
          case findPivot r c mat of
            Nothing -> go r (c + 1) mat rhs pivots
            Just p ->
              let (mat1, rhs1) = swapRows r p mat rhs
                  pivotVal = (mat1 !! r) !! c
                  (mat2, rhs2) = normalizeRow r c pivotVal mat1 rhs1
                  (mat3, rhs3) = eliminate r c mat2 rhs2
                  (rest, restRhs, ps) = go (r + 1) (c + 1) mat3 rhs3 (pivots ++ [c])
               in (rest, restRhs, ps)

    findPivot r c mat = L.find (\i -> (mat !! i) !! c /= 0) [r .. rows - 1]
    swapRows i j mat rhs =
      let mat' = swap i j mat
          rhs' = swap i j rhs
       in (mat', rhs')
    swap i j xs =
      let xi = xs !! i
          xj = xs !! j
       in [ if k == i then xj else if k == j then xi else xs !! k | k <- [0 .. length xs - 1] ]
    normalizeRow r c pivot mat rhs =
      let factor = 1 / pivot
          mat' =
            [ if i == r
                then [ if j < c then mat !! i !! j else mat !! i !! j * factor | j <- [0 .. cols - 1] ]
                else mat !! i
              | i <- [0 .. rows - 1]
            ]
          rhs' = [ if i == r then rhs !! i * factor else rhs !! i | i <- [0 .. rows - 1] ]
       in (mat', rhs')
    eliminate r c mat rhs =
      let mat' =
            [ if i == r
                then mat !! i
                else
                  let factor = mat !! i !! c
                   in [ mat !! i !! j - factor * (mat !! r !! j) | j <- [0 .. cols - 1] ]
              | i <- [0 .. rows - 1]
            ]
          rhs' =
            [ if i == r
                then rhs !! i
                else rhs !! i - (mat !! i !! c) * (rhs !! r)
              | i <- [0 .. rows - 1]
            ]
       in (mat', rhs')

-- solve square system, returning Nothing if singular
solveSquare :: [[Rational]] -> [Rational] -> Maybe [Rational]
solveSquare mat rhs =
  let n = length mat
      (mat', rhs', _) = gaussJordan mat rhs
      leadCols = map (L.findIndex (/= 0)) mat'
   in if any (== Nothing) leadCols
        then Nothing
        else
          let xs = replicate n 0
              sol =
                foldr
                  ( \(i, mrow, bval) acc ->
                      case L.findIndex (/= 0) mrow of
                        Nothing -> acc
                        Just c ->
                          let rest = sum [mrow !! j * acc !! j | j <- [c + 1 .. n - 1]]
                              val = bval - rest
                           in take c acc ++ [val] ++ drop (c + 1) acc
                  )
                  xs
                  (zip3 [0 ..] mat' rhs')
           in Just sol

combinations :: Int -> [a] -> [[a]]
combinations 0 _ = [[]]
combinations _ [] = []
combinations k (x : xs) = map (x :) (combinations (k - 1) xs) ++ combinations k xs

boundsForFree :: [([Rational], Rational)] -> Int -> ([Rational], [Rational])
boundsForFree constraints f =
  let subsets = combinations f [0 .. length constraints - 1]
      initMin = replicate f (Nothing :: Maybe Rational)
      initMax = replicate f (0 :: Rational)
      (mn, mx) =
        foldl
          ( \(mins, maxs) subset ->
              let mat = [fst (constraints !! i) | i <- subset]
                  rhs = [snd (constraints !! i) | i <- subset]
               in case solveSquare mat rhs of
                    Nothing -> (mins, maxs)
                    Just sol ->
                      if feasible sol constraints
                        then
                          ( zipWith (\m v -> Just $ maybe v (min v) m) mins sol,
                            zipWith max maxs sol
                          )
                        else (mins, maxs)
          )
          (initMin, initMax)
          subsets
   in (map (maybe 0 id) mn, mx)
  where
    feasible sol cons = all (\(c, b) -> dot c sol <= b) cons
    dot v w = sum (zipWith (*) v w)

minPressJolts :: [[Int]] -> [Int] -> Int
minPressJolts buttons target =
  let m = length target
      n = length buttons
      a = [[if i `elem` btn then 1 else 0 | btn <- buttons] | i <- [0 .. m - 1]]
      aRat = map (map fromIntegral) a
      bRat = map fromIntegral target
      (aRed, bRed, pivots) = gaussJordan aRat bRat
      free = [c | c <- [0 .. n - 1], c `notElem` pivots]
      fCount = length free
      rowInfo =
        [ (bRed !! r, [aRed !! r !! f | f <- free])
          | (r, p) <- zip [0 ..] pivots,
            let _ = p
        ]
   in if fCount == 0
        then
          let pivotVals = map fst rowInfo
           in if all (\x -> x >= 0 && denominatorIsOne x) pivotVals
                then sum (map round pivotVals)
                else error "no feasible solution"
        else
          let constraints =
                [ (coeff, b) | (b, coeff) <- rowInfo ]
                  ++ [ (map (negAt j) [0 .. fCount - 1], 0) | j <- [0 .. fCount - 1] ]
              (mins, maxs) = boundsForFree constraints fCount
              bounds =
                [ (max 0 (ceilingR m'), floorR x) | (m', x) <- zip mins maxs ]
           in search bounds rowInfo

negAt :: Int -> Int -> Rational
negAt j idx = if idx == j then (-1) else 0

denominatorIsOne :: Rational -> Bool
denominatorIsOne x = denominator x == 1

ceilingR :: Rational -> Int
ceilingR = ceiling . fromRational

floorR :: Rational -> Int
floorR = floor . fromRational

search :: [(Int, Int)] -> [(Rational, [Rational])] -> Int
search bounds rows = go 0 []
  where
    fCount = length bounds
    go idx vals
      | idx == fCount =
          let pivots = map (\(b, coeff) -> b - sum (zipWith (*) coeff (map fromIntegral vals))) rows
           in if all (\v -> v >= 0 && denominatorIsOne v) pivots
                then sum vals + sum (map round pivots)
                else maxBound `div` 4
      | otherwise =
          let (lo, hi) = bounds !! idx
           in foldl
                (\acc v -> min acc (go (idx + 1) (vals ++ [v])))
                (maxBound `div` 4)
                [lo .. hi]

main :: IO ()
main = do
  raw <- readFile "d10.txt"
  let machines = lines raw
      parsed = map parseLine machines

      part1 =
        sum
          [ let target = patternMask pat
                masks = map buttonMask btns
             in minPressLights target masks
            | (pat, btns, _) <- parsed
          ]

      part2 =
        sum
          [ minPressJolts btns targets
            | (_, btns, targets) <- parsed
          ]

  putStrLn $ "p1: " ++ show part1
  putStrLn $ "p2: " ++ show part2

File: content/puzzle/aoc/25/d10.rs
use regex::Regex;
use std::collections::{HashSet, VecDeque};
use std::fs;
use std::sync::LazyLock;

static PAT_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\[([.#]+)\]").unwrap());
static BTN_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\(([0-9,]+)\)").unwrap());
static TGT_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\{([0-9,]+)\}").unwrap());

type Rational = (i64, i64); // (numerator, denominator)

fn gcd(a: i64, b: i64) -> i64 {
    if b == 0 {
        a.abs()
    } else {
        gcd(b, a % b)
    }
}

fn rat_new(n: i64, d: i64) -> Rational {
    if d == 0 {
        panic!("division by zero");
    }
    let g = gcd(n, d);
    let sign = if d < 0 { -1 } else { 1 };
    (sign * n / g, sign * d / g)
}

fn rat_add(a: Rational, b: Rational) -> Rational {
    rat_new(a.0 * b.1 + b.0 * a.1, a.1 * b.1)
}

fn rat_sub(a: Rational, b: Rational) -> Rational {
    rat_new(a.0 * b.1 - b.0 * a.1, a.1 * b.1)
}

fn rat_mul(a: Rational, b: Rational) -> Rational {
    rat_new(a.0 * b.0, a.1 * b.1)
}

fn rat_div(a: Rational, b: Rational) -> Rational {
    rat_new(a.0 * b.1, a.1 * b.0)
}

fn rat_neg(a: Rational) -> Rational {
    (-a.0, a.1)
}
fn rat_zero() -> Rational {
    (0, 1)
}
fn rat_one() -> Rational {
    (1, 1)
}
fn rat_from(n: i64) -> Rational {
    (n, 1)
}
fn rat_is_zero(a: Rational) -> bool {
    a.0 == 0
}
fn rat_le(a: Rational, b: Rational) -> bool {
    a.0 * b.1 <= b.0 * a.1
}
fn rat_ge_zero(a: Rational) -> bool {
    a.0 >= 0
}
fn rat_is_int(a: Rational) -> bool {
    a.1 == 1 || a.0 % a.1 == 0
}
fn rat_to_int(a: Rational) -> i64 {
    a.0 / a.1
}

fn rat_floor(a: Rational) -> i64 {
    if a.0 >= 0 {
        a.0 / a.1
    } else {
        (a.0 - a.1 + 1) / a.1
    }
}

fn rat_ceil(a: Rational) -> i64 {
    if a.0 >= 0 {
        (a.0 + a.1 - 1) / a.1
    } else {
        a.0 / a.1
    }
}

fn parse_line(s: &str) -> (String, Vec<Vec<usize>>, Vec<i64>) {
    let pattern = PAT_RE.captures(s).unwrap()[1].to_string();

    let buttons: Vec<Vec<usize>> = BTN_RE
        .captures_iter(s)
        .map(|cap| {
            cap[1]
                .split(',')
                .map(|x| x.trim().parse().unwrap())
                .collect()
        })
        .collect();

    let targets: Vec<i64> = TGT_RE.captures(s).unwrap()[1]
        .split(',')
        .map(|x| x.trim().parse().unwrap())
        .collect();

    (pattern, buttons, targets)
}

// part 1: BFS over XOR states
fn pattern_to_mask(pat: &str) -> u64 {
    pat.chars()
        .enumerate()
        .fold(0, |acc, (i, c)| if c == '#' { acc | (1 << i) } else { acc })
}

fn button_to_mask(btn: &[usize]) -> u64 {
    btn.iter().fold(0, |acc, &i| acc ^ (1 << i))
}

fn min_press_lights(target: u64, masks: &[u64]) -> u32 {
    if target == 0 {
        return 0;
    }
    let mut seen = HashSet::new();
    seen.insert(0u64);
    let mut q = VecDeque::new();
    q.push_back((0u64, 0u32));

    while let Some((state, dist)) = q.pop_front() {
        for &m in masks {
            let nxt = state ^ m;
            if nxt == target {
                return dist + 1;
            }
            if !seen.contains(&nxt) {
                seen.insert(nxt);
                q.push_back((nxt, dist + 1));
            }
        }
    }
    panic!("unreachable");
}

// part 2: gaussian elimination
fn gauss_jordan(
    a: &[Vec<Rational>],
    b: &[Rational],
) -> (Vec<Vec<Rational>>, Vec<Rational>, Vec<usize>) {
    let rows = a.len();
    let cols = if rows > 0 { a[0].len() } else { 0 };
    let mut mat: Vec<Vec<Rational>> = a.to_vec();
    let mut rhs: Vec<Rational> = b.to_vec();
    let mut pivots = Vec::new();
    let mut r = 0;

    for c in 0..cols {
        let pivot_row = (r..rows).find(|&i| !rat_is_zero(mat[i][c]));
        let pr = match pivot_row {
            Some(p) => p,
            None => continue,
        };
        mat.swap(r, pr);
        rhs.swap(r, pr);

        let scale = mat[r][c];
        for j in 0..cols {
            mat[r][j] = rat_div(mat[r][j], scale);
        }
        rhs[r] = rat_div(rhs[r], scale);

        for i in 0..rows {
            if i != r && !rat_is_zero(mat[i][c]) {
                let factor = mat[i][c];
                for j in 0..cols {
                    mat[i][j] = rat_sub(mat[i][j], rat_mul(factor, mat[r][j]));
                }
                rhs[i] = rat_sub(rhs[i], rat_mul(factor, rhs[r]));
            }
        }
        pivots.push(c);
        r += 1;
        if r >= rows {
            break;
        }
    }
    (mat, rhs, pivots)
}

fn solve_square(mat: &[Vec<Rational>], rhs: &[Rational]) -> Option<Vec<Rational>> {
    let n = mat.len();
    let (mat2, rhs2, _) = gauss_jordan(mat, rhs);
    let mut sol = vec![rat_zero(); n];

    for i in (0..n).rev() {
        let lead = (0..n).find(|&j| !rat_is_zero(mat2[i][j]))?;
        let rest = (lead + 1..n).fold(rat_zero(), |acc, j| {
            rat_add(acc, rat_mul(mat2[i][j], sol[j]))
        });
        sol[lead] = rat_sub(rhs2[i], rest);
    }
    Some(sol)
}

fn combinations(n: usize, k: usize) -> Vec<Vec<usize>> {
    if k == 0 {
        return vec![vec![]];
    }
    if n < k {
        return vec![];
    }
    let mut result = Vec::new();
    fn gen(
        start: usize,
        k: usize,
        n: usize,
        current: &mut Vec<usize>,
        result: &mut Vec<Vec<usize>>,
    ) {
        if current.len() == k {
            result.push(current.clone());
            return;
        }
        for i in start..=(n - k + current.len()) {
            current.push(i);
            gen(i + 1, k, n, current, result);
            current.pop();
        }
    }
    gen(0, k, n, &mut Vec::new(), &mut result);
    result
}

fn bounds_for_free(
    constraints: &[(Vec<Rational>, Rational)],
    f: usize,
) -> (Vec<Rational>, Vec<Rational>) {
    let mut mins: Vec<Option<Rational>> = vec![None; f];
    let mut maxs = vec![rat_zero(); f];

    for subset in combinations(constraints.len(), f) {
        let mat: Vec<Vec<Rational>> = subset.iter().map(|&i| constraints[i].0.clone()).collect();
        let rhs: Vec<Rational> = subset.iter().map(|&i| constraints[i].1).collect();

        if let Some(sol) = solve_square(&mat, &rhs) {
            let feasible = constraints.iter().all(|(c, b)| {
                let dot = c
                    .iter()
                    .zip(&sol)
                    .fold(rat_zero(), |acc, (&ci, &si)| rat_add(acc, rat_mul(ci, si)));
                rat_le(dot, *b)
            });
            if feasible {
                for j in 0..f {
                    mins[j] = Some(match mins[j] {
                        None => sol[j],
                        Some(m) if rat_le(sol[j], m) => sol[j],
                        Some(m) => m,
                    });
                    if rat_le(maxs[j], sol[j]) {
                        maxs[j] = sol[j];
                    }
                }
            }
        }
    }
    (
        mins.into_iter().map(|m| m.unwrap_or(rat_zero())).collect(),
        maxs,
    )
}

fn min_press_jolts(buttons: &[Vec<usize>], targets: &[i64]) -> i64 {
    let m = targets.len();
    let n = buttons.len();

    let a: Vec<Vec<Rational>> = (0..m)
        .map(|i| {
            buttons
                .iter()
                .map(|btn| {
                    if btn.contains(&i) {
                        rat_one()
                    } else {
                        rat_zero()
                    }
                })
                .collect()
        })
        .collect();
    let b: Vec<Rational> = targets.iter().map(|&t| rat_from(t)).collect();

    let (a_red, b_red, pivots) = gauss_jordan(&a, &b);
    let free: Vec<usize> = (0..n).filter(|c| !pivots.contains(c)).collect();
    let f_count = free.len();

    if f_count == 0 {
        let pivot_vals: Vec<Rational> = (0..pivots.len()).map(|r| b_red[r]).collect();
        if pivot_vals.iter().all(|&v| rat_ge_zero(v) && rat_is_int(v)) {
            return pivot_vals.iter().map(|&v| rat_to_int(v)).sum();
        }
        panic!("no feasible solution");
    }

    let row_info: Vec<(Rational, Vec<Rational>)> = (0..pivots.len())
        .map(|r| (b_red[r], free.iter().map(|&f| a_red[r][f]).collect()))
        .collect();

    let mut constraints: Vec<(Vec<Rational>, Rational)> = row_info
        .iter()
        .map(|(b_val, coeff)| (coeff.clone(), *b_val))
        .collect();

    for j in 0..f_count {
        let neg: Vec<Rational> = (0..f_count)
            .map(|k| {
                if k == j {
                    rat_neg(rat_one())
                } else {
                    rat_zero()
                }
            })
            .collect();
        constraints.push((neg, rat_zero()));
    }

    let (mins, maxs) = bounds_for_free(&constraints, f_count);
    let bounds: Vec<(i64, i64)> = mins
        .iter()
        .zip(&maxs)
        .map(|(&mi, &mx)| (rat_ceil(mi).max(0), rat_floor(mx)))
        .collect();

    fn search(
        idx: usize,
        vals: &mut Vec<i64>,
        bounds: &[(i64, i64)],
        row_info: &[(Rational, Vec<Rational>)],
        f_count: usize,
    ) -> i64 {
        if idx == f_count {
            let pivot_vals: Vec<Rational> = row_info
                .iter()
                .map(|(b_val, coeff)| {
                    let sum = coeff
                        .iter()
                        .zip(vals.iter())
                        .fold(rat_zero(), |acc, (&c, &v)| {
                            rat_add(acc, rat_mul(c, rat_from(v)))
                        });
                    rat_sub(*b_val, sum)
                })
                .collect();
            if pivot_vals.iter().all(|&v| rat_ge_zero(v) && rat_is_int(v)) {
                return vals.iter().sum::<i64>()
                    + pivot_vals.iter().map(|&v| rat_to_int(v)).sum::<i64>();
            }
            return i64::MAX / 4;
        }
        let (lo, hi) = bounds[idx];
        let mut best = i64::MAX / 4;
        for v in lo..=hi {
            vals.push(v);
            best = best.min(search(idx + 1, vals, bounds, row_info, f_count));
            vals.pop();
        }
        best
    }

    search(0, &mut Vec::new(), &bounds, &row_info, f_count)
}

fn main() {
    let raw = fs::read_to_string("d10.txt").expect("cannot read d10.txt");
    let lines: Vec<&str> = raw.lines().filter(|l| !l.is_empty()).collect();

    let p1: u32 = lines
        .iter()
        .map(|ln| {
            let (pat, btns, _) = parse_line(ln);
            let target = pattern_to_mask(&pat);
            let masks: Vec<u64> = btns.iter().map(|b| button_to_mask(b)).collect();
            min_press_lights(target, &masks)
        })
        .sum();

    let p2: i64 = lines
        .iter()
        .map(|ln| {
            let (_, btns, targets) = parse_line(ln);
            min_press_jolts(&btns, &targets)
        })
        .sum();

    println!("p1: {}", p1);
    println!("p2: {}", p2);
}
